### PR TITLE
[api] Cache for BlockByHeight and TransactionLogs

### PR DESCRIPTION
**Pull request by envestcc**
# Description

as title

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by CoderZhi — COMMENTED** [review]
How about https://github.com/iotexproject/iotex-core/pull/4446/files#diff-cab2e4d938c21357ad6f17be5bad1c5c1672f85dbc3ee8493769b48f04658debR302-R316 ?

**Comment by envestcc** [comment]
> How about https://github.com/iotexproject/iotex-core/pull/4446/files#diff-cab2e4d938c21357ad6f17be5bad1c5c1672f85dbc3ee8493769b48f04658debR302-R316 ?

block doesn't contain receipts, so the `GetReceipts` may not use the block cache

**Review by CoderZhi — APPROVED** [review]

File: blockchain/blockdao/blockdao.go [hunk 1]
```diff
@@ -166,12 +172,22 @@ func (dao *blockDAO) Stop(ctx context.Context) error {
...
+		return v.(hash.Hash256), nil
+	}
+	_cacheMtc.WithLabelValues("miss_block").Inc()
 	timer := dao.timerFactory.NewTimer("get_block_hash")
 	defer timer.End()
 	return dao.blockStore.GetBlockHash(height)
 }
 
 func (dao *blockDAO) GetBlockHeight(hash hash.Hash256) (uint64, error) {
+	if blk, ok := lruCacheGet(dao.blockCache, hash); ok {
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:186]: also check `headerCache`

File: blockchain/blockdao/blockdao.go [hunk 2]
```diff
@@ -241,8 +285,11 @@ func (dao *blockDAO) Header(h hash.Hash256) (*block.Header, error) {
 		_cacheMtc.WithLabelValues("hit_header").Inc()
 		return header.(*block.Header), nil
 	}
+	if header, ok := lruCacheGet(dao.blockCache, h); ok {
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:288]: `blk, ok :=`

File: blockchain/blockdao/blockdao.go [hunk 3]
```diff
@@ -241,8 +285,11 @@ func (dao *blockDAO) Header(h hash.Hash256) (*block.Header, error) {
...
+		_cacheMtc.WithLabelValues("hit_block").Inc()
+		return &header.(*block.Block).Header, nil
+	}
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:291]: can add to `dao.heightToHash`?

File: blockchain/blockdao/blockdao.go [hunk 4]
```diff
@@ -166,12 +172,22 @@ func (dao *blockDAO) Stop(ctx context.Context) error {
 }
 
 func (dao *blockDAO) GetBlockHash(height uint64) (hash.Hash256, error) {
+	if v, ok := dao.heightToHash.Load(height); ok {
+		_cacheMtc.WithLabelValues("hit_block").Inc()
+		return v.(hash.Hash256), nil
+	}
+	_cacheMtc.WithLabelValues("miss_block").Inc()
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:179]: can update `dao.heightToHash`

File: blockchain/blockdao/blockdao.go [hunk 5]
```diff
@@ -191,22 +207,47 @@ func (dao *blockDAO) GetBlock(hash hash.Hash256) (*block.Block, error) {
 		return nil, err
 	}
 	lruCachePut(dao.blockCache, hash, blk)
+	dao.heightToHash.Store(blk.Height(), hash)
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:210]: can update `dao.headerCache`?

File: blockchain/blockdao/blockdao.go [hunk 6]
```diff
@@ -191,22 +207,47 @@ func (dao *blockDAO) GetBlock(hash hash.Hash256) (*block.Block, error) {
...
 	timer := dao.timerFactory.NewTimer("get_block_byheight")
 	defer timer.End()
-	return dao.blockStore.GetBlockByHeight(height)
+	blk, err := dao.blockStore.GetBlockByHeight(height)
+	if err != nil {
+		return nil, err
+	}
+	hash := blk.HashBlock()
+	lruCachePut(dao.blockCache, hash, blk)
+	dao.heightToHash.Store(blk.Height(), hash)
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:228]: can update `dao.headerCache`?

File: blockchain/blockdao/blockdao.go [hunk 7]
```diff
@@ -71,6 +72,8 @@ type (
 		footerCache  cache.LRUCache
 		receiptCache cache.LRUCache
 		blockCache   cache.LRUCache
+		txLogCache   cache.LRUCache
+		heightToHash sync.Map
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:76]: can avoid this by saving both Height() and Hash() to LRU

File: blockchain/blockdao/blockdao.go [hunk 8]
```diff
@@ -191,22 +212,43 @@ func (dao *blockDAO) GetBlock(hash hash.Hash256) (*block.Block, error) {
...
+	}
+	_cacheMtc.WithLabelValues("miss_block").Inc()
 	timer := dao.timerFactory.NewTimer("get_block_byheight")
 	defer timer.End()
-	return dao.blockStore.GetBlockByHeight(height)
+	blk, err := dao.blockStore.GetBlockByHeight(height)
+	if err != nil {
+		return nil, err
+	}
+	lruCachePut(dao.blockCache, height, blk)
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:231]: `lruCachePut(dao.blockCache, hash, blk)`

File: blockchain/blockdao/blockdao.go [hunk 9]
```diff
@@ -166,12 +168,31 @@ func (dao *blockDAO) Stop(ctx context.Context) error {
...
+		return block.GenesisHash(), nil
+	}
+	if v, ok := lruCacheGet(dao.headerCache, height); ok {
+		_cacheMtc.WithLabelValues("hit_header").Inc()
+		return v.(*block.Header).HashBlock(), nil
+	}
+	if v, ok := lruCacheGet(dao.blockCache, height); ok {
+		_cacheMtc.WithLabelValues("hit_block").Inc()
+		return v.(*block.Block).HashBlock(), nil
+	}
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:181]:
```
if v := dao.headerFromCache(height); v != nil {
		return v.HashHeader(), nil
}
```

**Review by dustinxie — APPROVED** [review]

**Pull request merged**